File: web/src/views/format.test.ts
import { describe, it, mock } from "node:test";
import assert from "node:assert/strict";
import { makeTimestamp } from "../test-helpers.ts";
import {
  timeAgo,
  formatTimestamp,
  scaleReasonLabel,
  functionKey,
  functionPath,
  durationBetween,
  eventTypeLabel,
  eventSeverityBadge,
} from "./format.ts";
import { timestampFromMs } from "@bufbuild/protobuf/wkt";

await describe("timeAgo", async () => {
  await it("returns dash for undefined", () => {
    assert.equal(timeAgo(undefined), "—");
  });

  await it("returns 'just now' for future timestamps", () => {
    const ts = makeTimestamp(60_000);
    assert.equal(timeAgo(ts), "just now");
  });

  await it("returns seconds for <60s", () => {
    mock.timers.enable({ apis: ["Date"] });
    const now = Date.now();
    mock.timers.setTime(now);

    const ts = makeTimestamp(-30_000);
    assert.equal(timeAgo(ts), "30s ago");

    mock.timers.reset();
  });

  await it("returns minutes for <60m", () => {
    mock.timers.enable({ apis: ["Date"] });
    const now = Date.now();
    mock.timers.setTime(now);

    const ts = makeTimestamp(-5 * 60_000);
    assert.equal(timeAgo(ts), "5m ago");

    mock.timers.reset();
  });

  await it("returns hours for <24h", () => {
    mock.timers.enable({ apis: ["Date"] });
    const now = Date.now();
    mock.timers.setTime(now);

    const ts = makeTimestamp(-3 * 3_600_000);
    assert.equal(timeAgo(ts), "3h ago");

    mock.timers.reset();
  });

  await it("returns days for >=24h", () => {
    mock.timers.enable({ apis: ["Date"] });
    const now = Date.now();
    mock.timers.setTime(now);

    const ts = makeTimestamp(-2 * 86_400_000);
    assert.equal(timeAgo(ts), "2d ago");

    mock.timers.reset();
  });

  await it("returns 0s ago for exactly now", () => {
    mock.timers.enable({ apis: ["Date"] });
    const now = Date.now();
    mock.timers.setTime(now);

    const ts = makeTimestamp(0);
    assert.equal(timeAgo(ts), "0s ago");

    mock.timers.reset();
  });
});

await describe("formatTimestamp", async () => {
  await it("returns dash for undefined", () => {
    assert.equal(formatTimestamp(undefined), "—");
  });

  await it("returns ISO string for a known timestamp", () => {
    const ts = makeTimestamp(0);
    const result = formatTimestamp(ts);
    // Should be a valid ISO 8601 string
    assert.match(result, /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}/);
  });
});

await describe("durationBetween", async () => {
  await it("returns dash for undefined start", () => {
    assert.equal(durationBetween(undefined, makeTimestamp(0)), "—");
  });

  await it("returns dash for undefined end", () => {
    assert.equal(durationBetween(makeTimestamp(0), undefined), "—");
  });

  await it("returns seconds for short durations", () => {
    const start = timestampFromMs(1000000);
    const end = timestampFromMs(1005000);
    assert.equal(durationBetween(start, end), "5s");
  });

  await it("returns minutes and seconds for longer durations", () => {
    const start = timestampFromMs(1000000);
    const end = timestampFromMs(1090000);
    assert.equal(durationBetween(start, end), "1m 30s");
  });

  await it("returns minutes only when no remaining seconds", () => {
    const start = timestampFromMs(1000000);
    const end = timestampFromMs(1120000);
    assert.equal(durationBetween(start, end), "2m");
  });

  await it("returns 0s for zero duration", () => {
    const ts = timestampFromMs(1000000);
    assert.equal(durationBetween(ts, ts), "0s");
  });
});

await describe("scaleReasonLabel", async () => {
  await it("maps known reason codes to labels", () => {
    assert.equal(scaleReasonLabel(0), "unspecified");
    assert.equal(scaleReasonLabel(1), "cpu");
    assert.equal(scaleReasonLabel(2), "heartbeat_timeout");
    assert.equal(scaleReasonLabel(3), "in_flight_requests");
    assert.equal(scaleReasonLabel(4), "memory");
    assert.equal(scaleReasonLabel(5), "no_ready_instances");
  });

  await it("returns unknown(N) for unrecognized codes", () => {
    assert.equal(scaleReasonLabel(6), "unknown(6)");
    assert.equal(scaleReasonLabel(99), "unknown(99)");
  });
});

await describe("eventTypeLabel", async () => {
  await it("maps known event type codes to labels", () => {
    assert.equal(eventTypeLabel(0), "unspecified");
    assert.equal(eventTypeLabel(1), "scale up");
    assert.equal(eventTypeLabel(2), "scale down");
    assert.equal(eventTypeLabel(3), "pod assigned");
    assert.equal(eventTypeLabel(4), "pod deleted");
    assert.equal(eventTypeLabel(5), "stale replacement");
    assert.equal(eventTypeLabel(6), "heartbeat timeout");
    assert.equal(eventTypeLabel(7), "stuck instance cleanup");
  });

  await it("returns unknown(N) for unrecognized codes", () => {
    assert.equal(eventTypeLabel(99), "unknown(99)");
  });
});

await describe("eventSeverityBadge", async () => {
  await it("returns info badge for severity 1", () => {
    assert.ok(eventSeverityBadge(1).includes("info"));
    assert.ok(eventSeverityBadge(1).includes("badge-muted"));
  });

  await it("returns warn badge for severity 2", () => {
    assert.ok(eventSeverityBadge(2).includes("warn"));
    assert.ok(eventSeverityBadge(2).includes("badge-yellow"));
  });

  await it("returns unknown badge for unrecognized severity", () => {
    assert.ok(eventSeverityBadge(0).includes("unknown"));
  });
});

await describe("functionKey", async () => {
  await it("joins namespace, deployment, and tenant with colons", () => {
    assert.equal(
      functionKey({ namespace: "prod", deployment: "api", tenant: "t1" }),
      "prod:api:t1",
    );
  });

  await it("handles empty strings", () => {
    assert.equal(functionKey({ namespace: "", deployment: "", tenant: "" }), "::");
  });
});

await describe("functionPath", async () => {
  await it("returns URL-encoded path", () => {
    assert.equal(
      functionPath({ namespace: "prod", deployment: "api", tenant: "t1" }),
      "/functions/prod%3Aapi%3At1",
    );
  });
});

File: internal/web/format_test.go
package web

import (
	"html/template"
	"testing"
	"time"

	"github.com/gadget-inc/skipper/internal/skipper"
	"google.golang.org/protobuf/types/known/timestamppb"
	"gotest.tools/v3/assert"
)

func TestTimeAgo(t *testing.T) {
	t.Parallel()

	testCases := []struct {
		name string
		ts   *timestamppb.Timestamp
		want string
	}{
		{name: "nil", ts: nil, want: "—"},
		{name: "just now", ts: timestamppb.Now(), want: "0s ago"},
		{name: "seconds", ts: timestamppb.New(time.Now().Add(-30 * time.Second)), want: "30s ago"},
		{name: "minutes", ts: timestamppb.New(time.Now().Add(-5 * time.Minute)), want: "5m ago"},
		{name: "hours", ts: timestamppb.New(time.Now().Add(-2 * time.Hour)), want: "2h ago"},
		{name: "days", ts: timestamppb.New(time.Now().Add(-48 * time.Hour)), want: "2d ago"},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			got := timeAgo(tc.ts)
			assert.Equal(t, got, tc.want)
		})
	}
}

func TestFormatTimestamp(t *testing.T) {
	t.Parallel()

	t.Run("nil", func(t *testing.T) {
		t.Parallel()
		assert.Equal(t, formatTimestamp(nil), "—")
	})

	t.Run("formats as RFC3339", func(t *testing.T) {
		t.Parallel()
		ts := timestamppb.New(time.Date(2025, 1, 15, 10, 30, 0, 0, time.UTC))
		assert.Equal(t, formatTimestamp(ts), "2025-01-15T10:30:00Z")
	})
}

func TestDurationBetween(t *testing.T) {
	t.Parallel()

	testCases := []struct {
		name  string
		start *timestamppb.Timestamp
		end   *timestamppb.Timestamp
		want  string
	}{
		{name: "nil start", start: nil, end: timestamppb.Now(), want: "—"},
		{name: "nil end", start: timestamppb.Now(), end: nil, want: "—"},
		{name: "seconds", start: timestamppb.New(time.Now()), end: timestamppb.New(time.Now().Add(15 * time.Second)), want: "15s"},
		{name: "minutes", start: timestamppb.New(time.Now()), end: timestamppb.New(time.Now().Add(90 * time.Second)), want: "1m 30s"},
		{name: "exact minutes", start: timestamppb.New(time.Now()), end: timestamppb.New(time.Now().Add(2 * time.Minute)), want: "2m"},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			got := durationBetween(tc.start, tc.end)
			assert.Equal(t, got, tc.want)
		})
	}
}

func TestFunctionKey(t *testing.T) {
	t.Parallel()

	fn := skipper.Assignment_builder{
		Namespace:  new("default"),
		Deployment: new("web-app"),
		Tenant:     new("tenant-1"),
	}.Build()

	assert.Equal(t, assignmentKey(fn), "default:web-app:tenant-1")
	assert.Equal(t, assignmentKey(nil), "")
}

func TestAssignmentPath(t *testing.T) {
	t.Parallel()

	fn := skipper.Assignment_builder{
		Namespace:  new("default"),
		Deployment: new("web-app"),
		Tenant:     new("tenant-1"),
	}.Build()

	assert.Equal(t, assignmentPath(fn), "/assignments/default:web-app:tenant-1")
}

func TestScaleReasonLabel(t *testing.T) {
	t.Parallel()

	assert.Equal(t, scaleReasonLabel(skipper.ScaleReason_SCALE_REASON_CPU), "cpu")
	assert.Equal(t, scaleReasonLabel(skipper.ScaleReason_SCALE_REASON_MEMORY), "memory")
	assert.Equal(t, scaleReasonLabel(skipper.ScaleReason_SCALE_REASON_HEARTBEAT_TIMEOUT), "heartbeat_timeout")
	assert.Equal(t, scaleReasonLabel(skipper.ScaleReason_SCALE_REASON_IN_FLIGHT_REQUESTS), "in_flight_requests")
	assert.Equal(t, scaleReasonLabel(skipper.ScaleReason_SCALE_REASON_NO_READY_INSTANCES), "no_ready_instances")
	assert.Equal(t, scaleReasonLabel(skipper.ScaleReason_SCALE_REASON_UNSPECIFIED), "unspecified")
}

func TestEventTypeLabel(t *testing.T) {
	t.Parallel()

	assert.Equal(t, eventTypeLabel(skipper.EventType_EVENT_TYPE_SCALE_UP), "scale up")
	assert.Equal(t, eventTypeLabel(skipper.EventType_EVENT_TYPE_SCALE_DOWN), "scale down")
	assert.Equal(t, eventTypeLabel(skipper.EventType_EVENT_TYPE_POD_ASSIGNED), "pod assigned")
	assert.Equal(t, eventTypeLabel(skipper.EventType_EVENT_TYPE_POD_DELETED), "pod deleted")
	assert.Equal(t, eventTypeLabel(skipper.EventType_EVENT_TYPE_STALE_REPLACEMENT), "stale replacement")
	assert.Equal(t, eventTypeLabel(skipper.EventType_EVENT_TYPE_HEARTBEAT_TIMEOUT), "heartbeat timeout")
	assert.Equal(t, eventTypeLabel(skipper.EventType_EVENT_TYPE_STUCK_INSTANCE_CLEANUP), "stuck instance cleanup")
	assert.Equal(t, eventTypeLabel(skipper.EventType_EVENT_TYPE_UNSPECIFIED), "unspecified")
}

func TestEventSeverityBadge(t *testing.T) {
	t.Parallel()

	testCases := []struct {
		name     string
		severity skipper.EventSeverity
		want     template.HTML
	}{
		{name: "warn", severity: skipper.EventSeverity_EVENT_SEVERITY_WARN, want: `<span class="badge badge-yellow">warn</span>`},
		{name: "info", severity: skipper.EventSeverity_EVENT_SEVERITY_INFO, want: `<span class="badge badge-muted">info</span>`},
		{name: "unspecified", severity: skipper.EventSeverity_EVENT_SEVERITY_UNSPECIFIED, want: `<span class="badge badge-muted">unknown</span>`},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			got := eventSeverityBadge(tc.severity)
			assert.Equal(t, string(got), string(tc.want))
		})
	}
}

func TestStatusBadge(t *testing.T) {
	t.Parallel()

	testCases := []struct {
		name  string
		ready bool
		want  template.HTML
	}{
		{name: "ready", ready: true, want: `<span class="badge badge-green">ready</span>`},
		{name: "pending", ready: false, want: `<span class="badge badge-yellow">pending</span>`},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			got := statusBadge(tc.ready)
			assert.Equal(t, string(got), string(tc.want))
		})
	}
}

func TestInstanceState(t *testing.T) {
	t.Parallel()

	testCases := []struct {
		name     string
		inst     *skipper.Instance
		activeRS string
		want     string
	}{
		{
			name: "ready instance on active RS",
			inst: skipper.Instance_builder{
				Name:       new("pod-1"),
				ReplicaSet: new("rs-1"),
				ReadyAt:    timestamppb.Now(),
			}.Build(),
			activeRS: "rs-1",
			want:     "ready",
		},
		{
			name: "pending instance on active RS",
			inst: skipper.Instance_builder{
				Name:       new("pod-2"),
				ReplicaSet: new("rs-1"),
			}.Build(),
			activeRS: "rs-1",
			want:     "pending",
		},
		{
			name: "ready instance on non-active RS",
			inst: skipper.Instance_builder{
				Name:       new("pod-3"),
				ReplicaSet: new("rs-old"),
				ReadyAt:    timestamppb.Now(),
			}.Build(),
			activeRS: "rs-1",
			want:     "stale",
		},
		{
			name: "ready instance with empty activeRS",
			inst: skipper.Instance_builder{
				Name:       new("pod-4"),
				ReplicaSet: new("rs-1"),
				ReadyAt:    timestamppb.Now(),
			}.Build(),
			activeRS: "",
			want:     "ready",
		},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			got := instanceState(tc.inst, tc.activeRS)
			assert.Equal(t, got, tc.want)
		})
	}
}

func TestInstanceStateBadge(t *testing.T) {
	t.Parallel()

	testCases := []struct {
		name     string
		inst     *skipper.Instance
		activeRS string
		want     template.HTML
	}{
		{
			name: "ready instance on active RS",
			inst: skipper.Instance_builder{
				Name:       new("pod-1"),
				ReplicaSet: new("rs-1"),
				ReadyAt:    timestamppb.Now(),
			}.Build(),
			activeRS: "rs-1",
			want:     `<span class="badge badge-green">ready</span>`,
		},
		{
			name: "pending instance on active RS",
			inst: skipper.Instance_builder{
				Name:       new("pod-2"),
				ReplicaSet: new("rs-1"),
			}.Build(),
			activeRS: "rs-1",
			want:     `<span class="badge badge-yellow">pending</span>`,
		},
		{
			name: "ready instance on non-active RS",
			inst: skipper.Instance_builder{
				Name:       new("pod-3"),
				ReplicaSet: new("rs-old"),
				ReadyAt:    timestamppb.Now(),
			}.Build(),
			activeRS: "rs-1",
			want:     `<span class="badge badge-yellow">stale</span>`,
		},
		{
			name: "ready instance with empty activeRS",
			inst: skipper.Instance_builder{
				Name:       new("pod-4"),
				ReplicaSet: new("rs-1"),
				ReadyAt:    timestamppb.Now(),
			}.Build(),
			activeRS: "",
			want:     `<span class="badge badge-green">ready</span>`,
		},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			got := instanceStateBadge(tc.inst, tc.activeRS)
			assert.Equal(t, string(got), string(tc.want))
		})
	}
}

func TestMetricsString(t *testing.T) {
	t.Parallel()

	testCases := []struct {
		name    string
		metrics []*skipper.ScaleMetric
		want    string
	}{
		{name: "nil", metrics: nil, want: "—"},
		{name: "empty", metrics: []*skipper.ScaleMetric{}, want: "—"},
		{
			name: "single metric",
			metrics: []*skipper.ScaleMetric{
				skipper.ScaleMetric_builder{Name: new("cpu"), Value: new(1.0)}.Build(),
			},
			want: "cpu: 1.0",
		},
		{
			name: "multiple metrics",
			metrics: []*skipper.ScaleMetric{
				skipper.ScaleMetric_builder{Name: new("cpu"), Value: new(50.0)}.Build(),
				skipper.ScaleMetric_builder{Name: new("memory"), Value: new(128.0)}.Build(),
			},
			want: "cpu: 50.0, memory: 128.0",
		},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			got := metricsString(tc.metrics)
			assert.Equal(t, got, tc.want)
		})
	}
}
